## Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# в первоначальном файле отсутствуют заголовки
# задаем названия заголоков на основе нумерации столбов в задании
# пропущенные данные ("?") заменим на NaN
columns = ['1_surgery?', '2', '3', '4_rectal_t', '5_pulse', '6', '7', '8', '9', '10', '11', '12', '13_abdominal_distension', '14', '15', '16', '17', '18', '19_packed_cell_volume', '20_total_protein', '21', '22', '23_outcome', '24_surgical_lesion?', '25', '26', '27', '28']
database = pd.read_csv('horse_data.csv', header=None, names=columns, na_values='?')
database

1_surgery?  2        3  4_rectal_t  5_pulse     6    7    8    9   10  \
0           2.0  1   530101        38.5     66.0  28.0  3.0  3.0  NaN  2.0   
1           1.0  1   534817        39.2     88.0  20.0  NaN  NaN  4.0  1.0   
2           2.0  1   530334        38.3     40.0  24.0  1.0  1.0  3.0  1.0   
3           1.0  9  5290409        39.1    164.0  84.0  4.0  1.0  6.0  2.0   
4           2.0  1   530255        37.3    104.0  35.0  NaN  NaN  6.0  2.0   
..          ... ..      ...         ...      ...   ...  ...  ...  ...  ...   
295         1.0  1   533886         NaN    120.0  70.0  4.0  NaN  4.0  2.0   
296         2.0  1   527702        37.2     72.0  24.0  3.0  2.0  4.0  2.0   
297         1.0  1   529386        37.5     72.0  30.0  4.0  3.0  4.0  1.0   
298         1.0  1   530612        36.5    100.0  24.0  3.0  3.0  3.0  1.0   
299         1.0  1   534618        37.2     40.0  20.0  NaN  NaN  NaN  NaN   

     ...  19_packed_cell_volume  20_total_protein   21   22  23_outcome  \
0    ...                   45.0               8.4  NaN  NaN         2.0   
1    ...                   50.0              85.0  2.0  2.0         3.0   
2    ...                   33.0               6.7  NaN  NaN         1.0   
3    ...                   48.0               7.2  3.0  5.3         2.0   
4    ...                   74.0               7.4  NaN  NaN         2.0   
..   ...                    ...               ...  ...  ...         ...   
295  ...                   55.0              65.0  NaN  NaN         3.0   
296  ...                   44.0               NaN  3.0  3.3         3.0   
297  ...                   60.0               6.8  NaN  NaN         2.0   
298  ...                   50.0               6.0  3.0  3.4         1.0   
299  ...                   36.0              62.0  1.0  1.0         3.0   

     24_surgical_lesion?     25  26  27  28  
0                      2  11300   0   0   2  
1                      2   2208   0   0   2  
2                      2      0   0   0   1  
3                      1   2208   0   0   1  
4                      2   4300   0   0   2  
..                   ...    ...  ..  ..  ..  
295                    2   3205   0   0   2  
296                    1   2208   0   0   1  
297                    1   3205   0   0   2  
298                    1   2208   0   0   1  
299                    2   6112   0   0   2  

[300 rows x 28 columns]

In [3]:
# выборка для анализа: 4 - количественные
sample_quantit = database.loc[:, ['4_rectal_t', '5_pulse', '19_packed_cell_volume', '20_total_protein']]
sample_quantit

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein
0          38.5     66.0                   45.0               8.4
1          39.2     88.0                   50.0              85.0
2          38.3     40.0                   33.0               6.7
3          39.1    164.0                   48.0               7.2
4          37.3    104.0                   74.0               7.4
..          ...      ...                    ...               ...
295         NaN    120.0                   55.0              65.0
296        37.2     72.0                   44.0               NaN
297        37.5     72.0                   60.0               6.8
298        36.5    100.0                   50.0               6.0
299        37.2     40.0                   36.0              62.0

[300 rows x 4 columns]

In [4]:
# выборка для анализа: 4 - качественные
sample_qualit = database.loc[:, ['1_surgery?', '13_abdominal_distension', '23_outcome', '24_surgical_lesion?']]
sample_qualit

1_surgery?  13_abdominal_distension  23_outcome  24_surgical_lesion?
0           2.0                      4.0         2.0                    2
1           1.0                      2.0         3.0                    2
2           2.0                      1.0         1.0                    2
3           1.0                      4.0         2.0                    1
4           2.0                      NaN         2.0                    2
..          ...                      ...         ...                  ...
295         1.0                      NaN         3.0                    2
296         2.0                      3.0         3.0                    1
297         1.0                      3.0         2.0                    1
298         1.0                      3.0         1.0                    1
299         1.0                      NaN         3.0                    2

[300 rows x 4 columns]

In [5]:
# расчет базовых метрик для количественных показателей
sample_quantit.describe()

4_rectal_t     5_pulse  19_packed_cell_volume  20_total_protein
count  240.000000  276.000000             271.000000        267.000000
mean    38.167917   71.913043              46.295203         24.456929
std      0.732289   28.630557              10.419335         27.475009
min     35.400000   30.000000              23.000000          3.300000
25%     37.800000   48.000000              38.000000          6.500000
50%     38.200000   64.000000              45.000000          7.500000
75%     38.500000   88.000000              52.000000         57.000000
max     40.800000  184.000000              75.000000         89.000000

##### Краткий комментарий для 4х столбцов с количестенными показателями
4_rectal_t:
 - Медиана и среднее близки, значит влияние выбросов не значительно, значения распределены нормально.
 - Учитывая, что нормальный уровень температуры - 37,8, данные в целом выглядят правдоподобно.

5_pulse:
 - Очень большой размах, максимальное значение выглядит слишком высоким при нормальном уровне в 30-40.
 - Довольно значительное влияние выбросов также отражает отклонение среднего от медианы.
 
19_packed_cell_volume:
 - Медиана и среднее близки, значит влияние выбросов не значительно, значения распределены нормально.
 - Большая часть лошадей из 300 не испытывала проблем с этим показателем (normal range is 30 to 50).
 
20_total_protein:
 - Размах и отклонение среднего от медианы позволяют сделать вывод о наличие значительных выбросов.
 - На первый взгляд есть некоторые сомнения в корректности данных, попадающих в выбросы (normal values lie in the 6-7.5 (gms/dL) range). Согласно описанию данных высокий уровень говорит об обезвоживании. Однако, уровень packed cell volume также может служить показателем обезвоживания, а его данные, на первый взгляд, не подтверждают значительную проблему с этим симптомом у лошадей.

In [6]:
# для качественных показателей рассчитаем моду, остальные метрики не информативны. 
sample_qualit.mode()

1_surgery?  13_abdominal_distension  23_outcome  24_surgical_lesion?
0         1.0                      1.0         1.0                    1

In [7]:
# доли уникальных значений для каждого из качественных показателей, так нагляднее
a = sample_qualit['1_surgery?'].value_counts(normalize=True)
b = sample_qualit['13_abdominal_distension'].value_counts(normalize=True)
c = sample_qualit['23_outcome'].value_counts(normalize=True)
d = sample_qualit['24_surgical_lesion?'].value_counts(normalize=True)

pd.concat([a, b, c, d], axis=1)

1_surgery?  13_abdominal_distension  23_outcome  24_surgical_lesion?
1.0    0.602007                 0.311475    0.595318             0.636667
2.0    0.397993                 0.266393    0.257525             0.363333
3.0         NaN                 0.266393    0.147157                  NaN
4.0         NaN                 0.155738         NaN                  NaN

##### Краткий комментарий для 4х столбцов с качестенными показателями
1_surgery?:
 - К основной части лошадей (60%) применяли хирургическое лечение.

13_abdominal_distension:
 - У большей части лошадей отсутстовал данный симптом. 
 
23_outcome:
 - Большая часть лошадей выжила.
 
24_surgical_lesion?:
 - Согласно ретроспектиной оценке основная часть проблем со здоровьем лошадей была хирургической.

## Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [8]:
# рассчет межквартильного размаха - 4_rectal_t

q1_4 = sample_quantit['4_rectal_t'].quantile(0.25)
q3_4 = sample_quantit['4_rectal_t'].quantile(0.75)
iqr_4 = q3_4 - q1_4
lower_bound_4 = q1_4 - (1.5*iqr_4)
higher_bound_4 = q3_4 + (1.5*iqr_4)
lower_outliners_4 = sample_quantit.loc[sample_quantit['4_rectal_t'].lt(lower_bound_4)]
higher_outliners_4 = sample_quantit.loc[sample_quantit['4_rectal_t'].gt(higher_bound_4)]
outliners_4 = pd.concat([lower_outliners_4, higher_outliners_4])
outliners_4

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein
44         35.4    140.0                   57.0              69.0
80         36.4     98.0                   47.0               6.4
118        36.5     78.0                   34.0              75.0
141        36.0     42.0                   64.0               6.8
238        36.1     88.0                   45.0               7.0
251        36.6     42.0                   52.0               7.1
298        36.5    100.0                   50.0               6.0
20         39.9     72.0                   46.0               6.1
54         40.3    114.0                   57.0               8.1
75         39.7    100.0                   48.0              57.0
91         40.3    114.0                   57.0               8.1
99         39.6    108.0                   59.0               8.0
259        40.8     72.0                   54.0               7.4
281        40.0     78.0                   66.0               6.5

In [9]:
# более подробне данные по лошади с минимальной температурой
database.loc[44, ['4_rectal_t', '5_pulse', '19_packed_cell_volume', '20_total_protein', '1_surgery?', '13_abdominal_distension', '23_outcome', '24_surgical_lesion?']]

4_rectal_t                  35.4
5_pulse                    140.0
19_packed_cell_volume       57.0
20_total_protein            69.0
1_surgery?                   1.0
13_abdominal_distension      NaN
23_outcome                   3.0
24_surgical_lesion?          1.0
Name: 44, dtype: float64

При анализе выбросов необходимо учитывать, что нормальная температура составляет 37.8. Повышенная температура может свидетельствовать о наличии инфекции. Температура тела ниже 37.8 может говорить о том, что животное прибывает в состоянии шока. Однако некоторые вопросы вызывает минимальная температура 35.4. Возможно ли такое в сочетании с довольно высоком уровне пульса в 140. Примем во внимание другие показатели по этой лошади. Судя по всему, лошадь была в крайне тяжелом состоянии, ее пришлось усыпить.

Все данные, попавшие в выбросы, оставляем в базе и продолжаем работу с ними, т.к. они не являются аномальными.

In [25]:
# рассчет межквартильного размаха - 5_pulse

q1_5 = sample_quantit['5_pulse'].quantile(0.25)
q3_5 = sample_quantit['5_pulse'].quantile(0.75)
iqr_5 = q3_5 - q1_5
lower_bound_5 = q1_5 - (1.5*iqr_5)
higher_bound_5 = q3_5 + (1.5*iqr_5)
lower_outliners_5 = sample_quantit.loc[sample_quantit['5_pulse'].lt(lower_bound_5)]
higher_outliners_5 = sample_quantit.loc[sample_quantit['5_pulse'].gt(higher_bound_5)]
outliners_5 = pd.concat([lower_outliners_5, higher_outliners_5])
outliners_5

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein
3          39.1    164.0                   48.0               7.2
41         39.0    150.0                   47.0               8.5
55         38.6    160.0                   38.0               NaN
255        38.8    184.0                   33.0               3.3
275        38.8    150.0                   50.0               6.2

In [27]:
# более подробне данные по лошадям
database.loc[[3, 41, 55, 255, 275], ['4_rectal_t', '5_pulse', '19_packed_cell_volume', '20_total_protein', '1_surgery?', '13_abdominal_distension', '23_outcome', '24_surgical_lesion?']]

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein  1_surgery?  \
3          39.1    164.0                   48.0               7.2         1.0   
41         39.0    150.0                   47.0               8.5         2.0   
55         38.6    160.0                   38.0               NaN         1.0   
255        38.8    184.0                   33.0               3.3         1.0   
275        38.8    150.0                   50.0               6.2         1.0   

     13_abdominal_distension  23_outcome  24_surgical_lesion?  
3                        4.0         2.0                    1  
41                       NaN         1.0                    1  
55                       4.0         2.0                    1  
255                      3.0         2.0                    1  
275                      2.0         2.0                    1

Нормальный уровень сердцебиения для взрослой лошади составляет 30-40 ударов/минуту. Высокое сердцебиение может свидетельствовать о том, что животные испытывали сильные боли или страдали от circulatory shock. У всех 5ти лошадей была повышена температура, но ниже уровней выбросов по температуре (проанализированы выше). Учитывая то, что 4 из 5 лошадей не выжили, они дейстительно были в плохом состоянии (хотя есть и другие гипотезы, например причиной смерти стало хирургическое вмешательство).

Выбросы значительно влияют на средний показатель и при этом сильно превышают нормальные значения. Прежде чем принять решение о включении/исключении выбросов, необходимо получить больше информации от профильных специалистов (ветеринаров).

In [28]:
# рассчет межквартильного размаха - 19_packed_cell_volume

q1_19 = sample_quantit['19_packed_cell_volume'].quantile(0.25)
q3_19 = sample_quantit['19_packed_cell_volume'].quantile(0.75)
iqr_19 = q3_19 - q1_19
lower_bound_19 = q1_19 - (1.5*iqr_19)
higher_bound_19 = q3_19 + (1.5*iqr_19)
lower_outliners_19 = sample_quantit.loc[sample_quantit['19_packed_cell_volume'].lt(lower_bound_19)]
higher_outliners_19 = sample_quantit.loc[sample_quantit['19_packed_cell_volume'].gt(higher_bound_19)]
outliners_19 = pd.concat([lower_outliners_19, higher_outliners_19])
outliners_19

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein
4          37.3    104.0                   74.0               7.4
144        37.1     84.0                   75.0              81.0
151        39.3      NaN                   75.0               NaN

In [29]:
# более подробне данные по лошадям
database.loc[[4, 144, 151], ['4_rectal_t', '5_pulse', '19_packed_cell_volume', '20_total_protein', '1_surgery?', '13_abdominal_distension', '23_outcome', '24_surgical_lesion?']]

4_rectal_t  5_pulse  19_packed_cell_volume  20_total_protein  1_surgery?  \
4          37.3    104.0                   74.0               7.4         2.0   
144        37.1     84.0                   75.0              81.0         1.0   
151        39.3      NaN                   75.0               NaN         1.0   

     13_abdominal_distension  23_outcome  24_surgical_lesion?  
4                        NaN         2.0                    2  
144                      4.0         3.0                    2  
151                      2.0         2.0                    1

В норме этот показатель варьируется от 30 до 50. Высокий уровень может быть вызван проблемами кровообращения либо обезвоживанием. Ни одна из этих лошадей не выжила, следовательно с некоторой долей уверенности можно говорить о том, что у них дейстительно были проблемы со здоровьем, которые могли сказаться на уровне анализируемого показателя.

Выявленные выбросы не оказывают значительного влияния на среднее. Не выглядят аномальными, принимаем к дальнейшему анализу.

In [18]:
# рассчет межквартильного размаха - 20_total_protein

q1_20 = sample_quantit['20_total_protein'].quantile(0.25)
q3_20 = sample_quantit['20_total_protein'].quantile(0.75)
iqr_20 = q3_20 - q1_20
lower_bound_20 = q1_20 - (1.5*iqr_20)
higher_bound_20 = q3_20 + (1.5*iqr_20)
lower_outliners_20 = sample_quantit['20_total_protein'].loc[sample_quantit['20_total_protein'].lt(lower_bound_20)]
higher_outliners_20 = sample_quantit['20_total_protein'].loc[sample_quantit['20_total_protein'].gt(higher_bound_20)]
outliners_20 = pd.concat([lower_outliners_20, higher_outliners_20])
outliners_20

Series([], Name: 20_total_protein, dtype: float64)

При расчете по выбранному критерию выбросы отсутствуют.
Норма находится на уровне 6 - 7.5. Медианное значение - 7.5, т.е. 50% животных имеют признаки обезвоживания. Крайне сложно судить о корректности данного показателя. Возможно необходимо использовать другой критерий для определения уровня выбросов. Нужно больше информации от профильных специалистов (ветеринаров).

## Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [13]:
# количество пропусков для выборки

selected_columns = ['4_rectal_t', '5_pulse', '19_packed_cell_volume', '20_total_protein', '1_surgery?', '13_abdominal_distension', '23_outcome', '24_surgical_lesion?']
database[selected_columns].isna().sum()

4_rectal_t                 60
5_pulse                    24
19_packed_cell_volume      29
20_total_protein           33
1_surgery?                  1
13_abdominal_distension    56
23_outcome                  1
24_surgical_lesion?         0
dtype: int64

##### 4_rectal_t:
20% (60/300) пропусков - по температуре. Это максимальное значение среди выбранных показателей.
Удаление или замена константой не является целесообразным для количественных показателей, т.к. может внести существенные искажения/затруднит расчеты. Как показал анализ в задании 1, среднее арифметическое и медиана приблизительно равны. Для целей заполнения пропусков возьмем медиану, рассчитанную по всему столбцу 4_rectal_t.

Результат замены представлен ниже:

In [19]:
db_fillna = pd.read_csv('horse_data.csv', header=None, names=columns, na_values='?')

db_fillna['4_rectal_t'].fillna(database['4_rectal_t'].median(), inplace=True)

print('Исходные данные:')
print(database['4_rectal_t'].describe())
print()
print('После замены:')
print(db_fillna['4_rectal_t'].describe())

Исходные данные:
count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: 4_rectal_t, dtype: float64

После замены:
count    300.000000
mean      38.174333
std        0.654831
min       35.400000
25%       37.900000
50%       38.200000
75%       38.500000
max       40.800000
Name: 4_rectal_t, dtype: float64


##### 5_pulse:
8% (24/300) пропусков в данной категории. Уровень относительно невысокий. Однако удаление или замена константой не является целесообразным для количественных показателей, т.к. повлияет/затруднит расчеты. Как показал анализ в задании 1, среднее арифметическое и медиана довольно значительно отличаются: 71.91 и 64 соответственно. Для целей заполнения пропусков возьмем медиану, как показатель наиболее устойчивый к выбросам. Результат замены представлен ниже:

In [23]:
db_fillna['5_pulse'].fillna(database['5_pulse'].median(), inplace=True)

print('Исходные данные:')
print(database['5_pulse'].describe())
print()
print('После замены:')
print(db_fillna['5_pulse'].describe())

Исходные данные:
count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: 5_pulse, dtype: float64

После замены:
count    300.000000
mean      71.280000
std       27.541545
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: 5_pulse, dtype: float64


##### 19_packed_cell_volume:
9.67% (29/300) пропусков в данной категории. Уровень относительно невысокий. Однако удаление или замена константой не является целесообразным для количественных показателей, т.к. повлияет/затруднит расчеты. Как показал анализ в задании 1, среднее арифметическое (46.3) и медиана (45) не слишком отличаются. Для целей заполнения пропусков возьмем медиану. Результат замены представлен ниже:

In [27]:
db_fillna['19_packed_cell_volume'].fillna(database['19_packed_cell_volume'].median(), inplace=True)

print('Исходные данные:')
print(database['19_packed_cell_volume'].describe())
print()
print('После замены:')
print(db_fillna['19_packed_cell_volume'].describe())

Исходные данные:
count    271.000000
mean      46.295203
std       10.419335
min       23.000000
25%       38.000000
50%       45.000000
75%       52.000000
max       75.000000
Name: 19_packed_cell_volume, dtype: float64

После замены:
count    300.000000
mean      46.170000
std        9.908584
min       23.000000
25%       39.000000
50%       45.000000
75%       50.000000
max       75.000000
Name: 19_packed_cell_volume, dtype: float64


##### 20_total_protein:
11% (33/300) пропусков. Уровень относительно невысокий. Однако удаление или замена константой не является целесообразным для количественных показателей, т.к. повлияет/затруднит расчеты. Как показал анализ в задании 1, среднее арифметическое (24.46) и медиана (7.5) значительно отличаются. В качестве показателя для заполнения пропусков возьмем медиану, как наименее чуствительный к выбросам показатель. Результаты замены ниже:

In [29]:
db_fillna['20_total_protein'].fillna(database['20_total_protein'].median(), inplace=True)

print('Исходные данные:')
print(database['20_total_protein'].describe())
print()
print('После замены:')
print(db_fillna['20_total_protein'].describe())

Исходные данные:
count    267.000000
mean      24.456929
std       27.475009
min        3.300000
25%        6.500000
50%        7.500000
75%       57.000000
max       89.000000
Name: 20_total_protein, dtype: float64

После замены:
count    300.000000
mean      22.591667
std       26.453849
min        3.300000
25%        6.600000
50%        7.500000
75%       53.250000
max       89.000000
Name: 20_total_protein, dtype: float64


##### 1_surgery?:
В данной категории всего 1 пропуск. Категория является качественной. Наиболее подходящей заменой является константа (no info), т.к. таблице предсталена медицинская информация (в данном случае: наличие или отсутствие операции) и будет нагляднее, если пропуски по качественным категориям будут выделены отдельно. Результат замены ниже:

In [30]:
db_fillna['1_surgery?'].fillna('no_info', inplace=True)

print('Исходные данные:')
print(database['1_surgery?'].value_counts())
print()
print('После замены:')
print(db_fillna['1_surgery?'].value_counts())

Исходные данные:
1.0    180
2.0    119
Name: 1_surgery?, dtype: int64

После замены:
1.0        180
2.0        119
no_info      1
Name: 1_surgery?, dtype: int64


##### 13_abdominal_distension:
Доля пропусков довольно высока - 18,7%. Согласно описанию параметр является важным, на его основе в том числе принимается решение о хирургическом вмешательстве. В качестве замены будет использована константа (no info), как наиболее адекватная для дальнейшей аналитики. Результат замены ниже:

In [32]:
db_fillna['13_abdominal_distension'].fillna('no_info', inplace=True)

print('Исходные данные:')
print(database['13_abdominal_distension'].value_counts())
print()
print('После замены:')
print(db_fillna['13_abdominal_distension'].value_counts())

Исходные данные:
1.0    76
3.0    65
2.0    65
4.0    38
Name: 13_abdominal_distension, dtype: int64

После замены:
1.0        76
3.0        65
2.0        65
no_info    56
4.0        38
Name: 13_abdominal_distension, dtype: int64


##### 23_outcome:
В данной категории всего 1 пропуск. Категория является качественной. Наиболее подходящей заменой является константа (no info), т.к. таблице предсталена медицинская информация (в данном случае: выжила ли лошадь) и будет нагляднее, если пропуски по качественным категориям будут выделены отдельно. Результат замены ниже:

In [33]:
db_fillna['23_outcome'].fillna('no_info', inplace=True)

print('Исходные данные:')
print(database['23_outcome'].value_counts())
print()
print('После замены:')
print(db_fillna['23_outcome'].value_counts())

Исходные данные:
1.0    178
2.0     77
3.0     44
Name: 23_outcome, dtype: int64

После замены:
1.0        178
2.0         77
3.0         44
no_info      1
Name: 23_outcome, dtype: int64


##### 24_surgical_lesion?:
Нет пропущенных данных.

In [34]:
# итоговый датафрейм без пропусков
db_fillna[selected_columns].isna().sum()

4_rectal_t                 0
5_pulse                    0
19_packed_cell_volume      0
20_total_protein           0
1_surgery?                 0
13_abdominal_distension    0
23_outcome                 0
24_surgical_lesion?        0
dtype: int64